pennylane/qchem/tapering.py

```diff
@@ -616,65 +781,47 @@ def taper_operation(operation, generators, paulixops, paulix_sector, wire_order,
-        mat1 = scipy.linalg.expm(1j * qml.matrix(gen_op, wire_order=wire_order) * coeffs)
-        mat2 = qml.matrix(operation, wire_order=wire_order)
-        phase = np.divide(mat1, mat2, out=np.zeros_like(mat1, dtype=complex), where=mat1 != 0)[
-            np.nonzero(np.round(mat1, 10))
-        ]
-        if not np.allclose(phase / phase[0], np.ones(len(phase))):  # check if the phase is global
-            raise ValueError(
-                f"Given gen_op: {gen_op} doesn't seem to be the correct generator for the {operation}."
-            )
-
+    # maintain a flag to track functional form of the operation
+    callable_op = callable(operation)
+    # get dummy objects in case functional form of operation or op_gen is being used
+    operation, op_gen = _build_callables(operation, op_wires=op_wires, op_gen=op_gen)
+
+    # build generator for the operation either internally or using the provided op_gen
+    op_gen = _build_generator(operation, wire_order, op_gen=op_gen)
+    # check compatibility between the generator and the symmeteries
     if np.all(
         [
             [
                 qml.is_commuting(op1, op2)
-                for op1, op2 in itertools.product(generator.ops, gen_op.ops)
+                for op1, op2 in itertools.product(generator.ops, op_gen.ops)
             ]
             for generator in generators
         ]
-    ):
-        gen_tapered = qml.taper(gen_op, generators, paulixops, paulix_sector)
+    ) and not np.all(np.isclose(op_gen.coeffs, np.zeros_like(op_gen.coeffs), rtol=1e-8)):
+        gen_tapered = qml.taper(op_gen, generators, paulixops, paulix_sector)
     else:
         gen_tapered = qml.Hamiltonian([], [])
     qml.simplify(gen_tapered)
 
-    params = operation.parameters[0] if len(operation.parameters) else 1.0
-    if qml.QueuingManager.recording():
-        qml.QueuingManager.update_info(operation, owner=gen_tapered)
-        for coeff, op in zip(*gen_tapered.terms()):
-            qml.PauliRot(-2 * params * coeff, qml.grouping.pauli_word_to_string(op), op.wires)
-    else:
-        ops_tapered = []
-        for coeff, op in zip(*gen_tapered.terms()):
-            ops_tapered.append(
-                qml.PauliRot(-2 * params * coeff, qml.grouping.pauli_word_to_string(op), op.wires)
-            )
-        return ops_tapered
+    def _tapered_op(params):
+        r"""Applies the tapered operation for the specified parameter value whenever
+        queing context is active, otherwise returns it as a list."""
+        if qml.QueuingManager.recording():
+            qml.QueuingManager.update_info(operation, owner=gen_tapered)
+            for coeff, op in zip(*gen_tapered.terms()):
+                qml.exp(op, 1j * params * coeff)
+        else:
+            ops_tapered = []
+            for coeff, op in zip(*gen_tapered.terms()):
+                ops_tapered.append(qml.exp(op, 1j * params * coeff))
+            return ops_tapered
```

Comment:
maybe add a test here when this function is called in a queuing context vs outside of one